site/_posts/2020-05-20-building-protected-api-with-rails-and-jwt.md

```diff
@@ -0,0 +1,254 @@
+```
+
+We use the HMAC signature algorithm because in this tutorial we control both the issuer of the token and the consumer--our API. We can therefore share a secret reliably between them. If we didn't have a good way to share secrets, using an asymmetric key would be a wiser choice.
+
+For this tutorial, we put the HMAC secret in the environment configuration files. For production usage, use your normal secrets management solution. You should make the HMAC secret a long string, but don't use the secret key base.
+
+Let's add our authorization code now that our tests fail because they are expecting certain unauthorized requests to return `:forbidden`.
+
+There are two places we could put the code that checks the token. We could add it to the `Messages` controller. Or we could add it to the `Application` controller. This latter choice would enforce authorization for all requests. Since we are building an API that should never be accessed without authorization, we should protect all our resources. If and when we need to distinguish between different claims (for instance, we may want to have some APIs only accessible for users with the `ADMIN` role), we can refactor and move the verification code to different controllers.
+
+Here's the authorization code for the `app/controllers/application_controller.rb` file:
+
+```ruby
+class ApplicationController < ActionController::API
+  before_action :require_jwt
+
+  def require_jwt
+    token = request.headers["HTTP_AUTHORIZATION"]
+    if !token
+      head :forbidden
+    end
+    if !valid_token(token)
+      head :forbidden
+    end
+  end
+
+  private
+  def valid_token(token)
+    unless token
+      return false
+    end
+
+    token.gsub!('Bearer ','')
+    begin
+      decoded_token = JWT.decode token, Rails.configuration.x.oauth.jwt_secret, true
+      return true
+    rescue JWT::DecodeError
+      Rails.logger.warn "Error decoding the JWT: "+ e.to_s
+    end
+    false
+  end
+end
+```
+
+We look for the JWT in the `AUTHORIZATION` HTTP header. If it doesn't exist, we deny access. If it does, we try to decode it. If it decodes without raising an exception, it is a valid JWT.
+
+## Verify claims
+
+But really, what does valid mean? That's something you define on an application by application basis. The `jwt` gem provides a baseline of default validation: [it checks the `exp` and `nbf` claims](https://github.com/jwt/ruby-jwt/blob/master/lib/jwt/default_options.rb), as well as verifying the signature. 
+
+But for this application, we need to be extra sure. After all, if our messages fell into the wrong hands, who knows what could happen?
+
+So let's verify additional claims are as they should be when we are decoding the JWT. We can do this by providing options to the `JWT.decode` method. Instead of:
```

Comment:
I think the first sentence here is missing a word or two and might have a few extras. Maybe it should be "So, let's verify that the values of the additional claims are what we expect when we are decoding the JWT."?

Reply:
👍